Sprint 12, intermediate solution for p2p.

uncle-sly opened the pull request:
Cхема DB, примеры SQL запросов согласно бизнес-логике. Таблицы схемы созданы, запросы запускал. Ошибки тем не менее, возможны.

u-sly committed:
Sprint 12, intermediate solution for p2p.

Cхема DB, примеры SQL запросов согласно бизнес-логике. Таблицы схемы созданы, запросы запускал. Ошибки тем не менее, возможны.

iolinpic commented on README.md:
По схеме:
- таблица названа film_category -а связь с genre? - обычно логика именования промежуточных таблиц, [имя первой сущности]_[имя второй сущности]
- primary_key id для промежуточной сущности, требуется только если ее собираются использовать как отдельную полноценную сущность - что в данном случае не предвидится.
- created_at поля - вроде как не требовались и нигде не используются, так что я бы предложил их убрать
- film_rating - изза перечисления через _ воспринимается не как таблица с рейтингом а как  связующая таблица (можно назвать просто rating)

uncle-sly replied:
* Ок, на film_genre поправил.  
* т.е. не обязательно так делать чтобы в каждой таблице был уникальный ключ? 
   для промежуточных можно и без?
* Ок, created_at для учебного проекта не пригодится наверно?
   но в целом вроде норм практика такое поле иметь как маркер временной?
* Ок

iolinpic replied:
- ну в целом можно сделать составной ключ, как совсем правильно думаю только те кто имеют много практики сказать смогут точно.
- created_at, и как и прочие поля данных должны появляться только если они востребованны - тоесть есть конечно места где их вообще в каждую таблицу пихают, есть те где нет, по факту любые данные ценны только если их используют. если например нужно в перспективе делать какието выборки по дате создания - например для блогов и новостей актуально

uncle-sly replied:
понятно. согласен.

iolinpic commented on README.md:
````diff
@@ -1,2 +1,80 @@
+>#### топ N наиболее популярных фильмов
+```sql
+SELECT f.name film_name,
+count(f.id) likes
````
скорее всего тут надо будет count(fl.user_id)

uncle-sly replied:
результат не поменяется вроде, но логичнее на (fl.user_id) по смыслу, чтобы читать удобнее?

iolinpic replied:
результат - да - но тут принцип такой же как в именовании переменных - если есть возможность взять такое поле, вычисления на котором будут понятны по смыслу, лучше его и взять. - ну и принцип наименования полей. Такие вещи начинаешь ценить только после того как походишь по граблям- вникая в чужие проекты, или в далеко забытые свои))

uncle-sly replied:
это отличный совет, полностью согласен)
спасибо, что заметил.

iolinpic commented on README.md:
````diff
@@ -1,2 +1,80 @@
+```
+
+---
+>#### список общих друзей с другим пользователем
````
подозреваю что можно проще - выбираем юзера и френда, лефтджоиним к ним  user_friend 2 раза по юзер и френд ид, группируем по ид юзеров и считаем где френдов больше 1

uncle-sly replied:
не показалось проще)
мысль в том, что сначала user с user_friend по id=user_id - юзеры и их друзья со стороны юзер, затем зеркально как бы, лефт джойним результат с user_friend по id=friend_id со стороны друзей, получаем общий резалтсет с друзьями тех и других. ну а дальше уже через having те которые есть общие со стороны юзера и френда. И исключаем самих юзера и френда.
правильно тебя понял ?

select uf1.user_id, uf2.friend_id, u."name" 
from public."user" u
join public.user_friends uf1 on u.id = uf1.user_id 
join public.user_friends uf2 on u.id = uf2.friend_id 
where u.id not in (1,3)
group by uf1.user_id, uf2.friend_id, u."name"
having count(uf2.friend_id) > 1

u-sly committed:
Sprint 12, intermediate solution for p2p.

Cхема DB, примеры SQL запросов согласно бизнес-логике. Таблицы схемы созданы, запросы запускал. Ошибки тем не менее, возможны.

TyutterinYakov reviewed (approved):
Апрув)
